ada/ast.py

```diff
@@ -12055,12 +12058,46 @@ def find_all_references(units=AnalysisUnit.array,
             dn.basic_decl.root_subp_declarations._or(dn.basic_decl.singleton)
         ))
 
+        # Determine if the entity is local. For the moment we only look at
+        # whether the entity is defined in a body. We might be able to do
+        # something for private entities as well, but it's much harder, and
+        # might not be as useful anyway.
+        is_local = Var(
+            dn.unit.root.cast(T.CompilationUnit).then(
+                lambda cu: And(
+                    cu.unit_kind == UnitBody,
+                    Not(cu.decl.as_bare_entity.defining_name
+                        == dn.node.as_bare_entity),
+                ),
+                default_val=False
+            )
+        )
+
         all_units = Var(bases.mapcat(
             lambda base: base.filter_is_imported_by(units, True)
         ).unique)
 
-        refs = Var(origin.bind(Self, all_units.mapcat(lambda u: u.root.then(
-            lambda r: dn.find_refs(r.as_bare_entity)
+        # Set of all roots to look into. Here, we try to optimize the set of
+        # roots we look into, with the following heuristics:
+        all_roots = Var(Cond(
+            # If the entity we're looking for is a parameter specification,
+            # then only look in the spec & the body of the subprogram.
+            dn.basic_decl.is_a(T.ParamSpec),
+            [dn.basic_decl.semantic_parent,
+             dn.next_part.basic_decl.semantic_parent],
```

Comment:
Unless I'm missing something, I think this is too constraining, as we lose references done through calls with named parameters with can be made from external units, e.g:
```
Foo (X => 2);
     ^ here
```